libtransact/src/context/execution_context.rs

```diff
@@ -0,0 +1,199 @@
+use crate::context::ContextId;
+use crate::receipts::Event;
+use std::fmt::Debug;
+use uuid::Uuid;
+
+use crate::receipts::StateChange;
+
+#[derive(Debug, Clone, Default)]
+pub struct Context<K, V> {
+    base_contexts: Vec<ContextId>,
+    state_changes: Vec<StateChange<K, V>>,
+    id: ContextId,
+    data: Vec<Vec<u8>>,
+    events: Vec<Event>,
+    state_id: String,
+}
+
+impl<K: PartialEq + Clone + Debug, V: Clone + Debug> Context<K, V> {
+    pub fn new(state_id: &str, base_contexts: Vec<ContextId>) -> Self {
+        Context {
+            base_contexts,
+            state_id: state_id.to_string(),
+            state_changes: Vec::new(),
+            id: *Uuid::new_v4().as_bytes(),
+            data: Vec::new(),
+            events: Vec::new(),
+        }
+    }
+
+    pub fn base_contexts(&self) -> &[ContextId] {
+        &self.base_contexts
+    }
+    pub fn events(&self) -> &Vec<Event> {
+        &self.events
+    }
+
+    pub fn state_changes(&self) -> &Vec<StateChange<K, V>> {
+        &self.state_changes
+    }
+
+    pub fn id(&self) -> &ContextId {
+        &self.id
+    }
+
+    pub fn data(&self) -> &Vec<Vec<u8>> {
+        &self.data
+    }
+
+    pub fn state_id(&self) -> &String {
+        &self.state_id
+    }
+
+    pub fn add_event(&mut self, event: Event) {
+        if !self.events().contains(&event) {
+            self.events.push(event);
+        }
+    }
+
+    pub fn add_data(&mut self, data: Vec<u8>) {
+        if !self.data().contains(&data) {
+            self.data.push(data);
+        }
+    }
+
+    /// Adds StateChange::Set without deleting previous StateChanges associated with the Key
+    pub fn set_state(&mut self, key: K, value: V) {
+        let new_state_change = StateChange::Set {
+            key: key.clone(),
+            value: value.clone(),
+        };
+        self.state_changes.push(new_state_change);
+    }
+
+    /// Adds StateChange::Delete and returns the value associated to the key being deleted
+    pub fn delete_state(&mut self, key: K) -> (K, Option<V>) {
+        let mut deleted_value: Option<V> = None;
+        let found_state_change = self
+            .state_changes()
+            .iter()
```

Comment:
Perhaps `iter_mut` should be used here, so we can avoid the clone.  Then we can just replace the state change, if we find it.  If not, we can push a new delete.

Reply:
We really just need the value associated with the state change, so I don't really need the clone of the entire list, but a clone of that value so that the manager can use it to return the value that was being deleted. If we just want to avoid cloning entirely, I can switch it to the behavior you suggested. Although, before when I had discussed this with Boyd, he had thought it might be nicer to have the entire history of state changes available.

Reply:
I don't that saving the history is necessary, since these structs aren't accessible outside of the context manager, and all the changes are collapsed into a final set state change anyway, with any intermediates dropped.